### ISSUE TITLE
compare() reports attribute differences even with check.attributes=FALSE

### ISSUE THREAD

**Issue by kenahoo** [hadley/testthat#519]
Here's an example:

```
> x <- 1:5; y <- x+0.2; attr(x, "thing") <- "value"
> compare(x, y, check.attributes=FALSE)
Attributes: < Modes: list, NULL >
Attributes: < Lengths: 1, 0 >
Attributes: < names for target but not for current >
Attributes: < current is not list-like >
```

Internally, this is because `compare()` (and thus all the functions that use it, like `expect_that()` and friends) first calls `all.equal()` to determine whether the comparison passes or fails - but then if it fails, it tries to reconstruct why it failed, but does so incorrectly.

One fix would be to check whether `list(...)` contains `check.attributes` with a `FALSE` value, and if so, skip the attribute reporting.

**Comment by richierocks** [comment]
I just fell over the same problem.

If I want to compare a named numeric vector with an unnamed numeric vector, then I should use `expect_equivalent`. Currently, the function sometimes erroneously complains about mismatched names.

```{r}
x <- setNames(1:5, letters[1:5])
y <- jitter(1:5)
expect_equivalent(x, y)
## Error: `x` not equivalent to `y`.
## names for target but not for current
```

I think the fix may be as simple as changing

```{r}
if (!same_attr(x, y)) {
  return(diff_attr(x, y))
}
```

to

```{r}
if (check.attributes && !same_attr(x, y)) {
  return(diff_attr(x, y))
}
```

in the `compare.*` methods.  (You'll probably need to make `check.attributes` one of the formal args too.)  I've not had chance to test this yet though.

**Comment by hadley** [comment]
Can you please provide a reprex that I can easily copy and paste into R?

**Comment by hadley** [comment]
Looks like this has already been fixed.

**Comment by kenahoo** [comment]
Fixed in cd90427 maybe?
